saml/src/main/java/com/linecorp/armeria/server/saml/HttpPostBindingUtil.java

```diff
@@ -0,0 +1,127 @@
+ */
+final class HttpPostBindingUtil {
+
+    private static final String[] XHTML = {
+            // 0
+            "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" +
+            "<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Strict//EN\" \"DTD/xhtml1-strict.dtd\">" +
+            "<html xmlns=\"http://www.w3.org/1999/xhtml\" xml:lang=\"en\" lang=\"en\">" +
+            "<head><meta http-equiv=\"content-type\" content=\"text/html;charset=utf-8\" /></head>" +
+            "<body onload=\"document.forms[0].submit()\">",
+            // 1, 2
+            "<form method=\"post\" action=\"", "\">",
+            // 3, 4, 5
+            "<input type=\"hidden\" name=\"", "\" value=\"", "\" />",
+            // 6
+            "</form></body></html>"
+    };
+
+    /**
+     * Returns an {@link HttpData} which holds a XHTML content.
+     */
+    static HttpData getHtmlForm(String remoteEndpointUrl,
+                                String paramName, String paramValue,
+                                @Nullable String relayState) {
+        requireNonNull(remoteEndpointUrl, "remoteEndpointUrl");
+        requireNonNull(paramName, "paramName");
+        requireNonNull(paramValue, "paramValue");
+
+        final StringBuilder html = new StringBuilder();
+        html.append(XHTML[0])
+            .append(XHTML[1]).append(StringEscapeUtils.escapeHtml(remoteEndpointUrl)).append(XHTML[2])
+            .append(XHTML[3]).append(StringEscapeUtils.escapeHtml(paramName))
+            .append(XHTML[4]).append(StringEscapeUtils.escapeHtml(paramValue)).append(XHTML[5]);
+
+        if (relayState != null) {
+            html.append(XHTML[3]).append(RELAY_STATE)
+                .append(XHTML[4]).append(StringEscapeUtils.escapeHtml(relayState)).append(XHTML[5]);
+        }
+        html.append(XHTML[6]);
+
+        return HttpData.ofUtf8(html.toString());
+    }
+
+    /**
+     * Signs the specified {@link SignableSAMLObject} with the specified {@link Credential} and
+     * {@code signatureAlgorithm}, and then encodes the object into a base64 string.
+     */
+    static String toSignedBase64(SignableSAMLObject signableObj,
+                                 Credential signingCredential,
+                                 String signatureAlgorithm) throws SamlException {
+        sign(signableObj, signingCredential, signatureAlgorithm);
+        final String messageStr = nodeToString(marshalMessage(signableObj));
+        return Base64.getEncoder().encodeToString(messageStr.getBytes(StandardCharsets.UTF_8));
+    }
+
+    /**
+     * Converts an {@link AggregatedHttpMessage} which is received from the remote entity to
+     * a {@link SAMLObject}.
+     */
+    static <T extends SAMLObject> MessageContext<T> toSamlObject(AggregatedHttpMessage msg,
+                                                                 String name) throws SamlException {
+        final Map<String, List<String>> parameters = SamlService.getParameters(msg);
```

Comment:
I'm not sure but otherwise we can define a static utility to replace parameters.get that returns the nullable value instead of list, to avoid the same null or empty check over and over.

Actually I've written a similar function in several codebases I've added armeria to, having it in core might be nice :)

Reply:
I added a simple wrapper class which can resolve what you pointed out. :-)

https://github.com/line/armeria/pull/1292/files#diff-6cb4c52b1669f6f34a67773b653f3585R160

We may have a better utility class in `core` in the future. At that time, we can replace the current implementation with new one. :-)